README.md
# Drftpd2Stable
 Distributed FTP Daemon

## Installation steps
DrFTPD 2 installation requires a number of steps before you can utilize the software to its full extend.
To give an overview of the installation process the different steps are listed below in this section.


* On the master you will need to:
  * Install Sun JAVA 1.6 or higher (JDK)
  * Install ANT or Eclipse on the master
  * Add needed plugins that are not present
  * Compile the software using setup wizard
  * Rename .dist files to .conf
  * Configure .conf files
* On the slaves you will need to:
  * Install Sun JAVA 1.6 or higher (JRE)
  * Copy slave.zip to a slave from the master
  * Configure slave.conf


## Install java
* Download and install a java development kit 6 (JDK) on the master.
* Download and install a java runtime environment 6 (JRE) or java development kit 6 (JDK) on the slaves.

`
You can get sun's JDK here: http://www.oracle.com/technetwork/java/javase/downloads/index.html

If you want to utilize blowfish in your environment also download Java Cryptography Extension (JCE) Unlimited Strength Jurisdiction Policy Files 6.
You will need to manually replace the files local_policy.jar and US_export_policy.jar in your java/jre/lib/security folder.

## Problems
* Ensure that JAVA_HOME is configured. You can check this using
  * Windows:"echo %JAVA_HOME%"
  * NiX: "echo $JAVA_HOME"

*TiP: Don't use Sun JAVA versions between 1.6_02 and 1.6_12. They will cause an error after hours of usage: java.net.SocketException: Too many open files
These are issues with your Operating System/Java Install and not related to DrFTPD.

## Install ant
Compiling DrFTPD is required to use the software.
To allow you to compile java you will need to install ANT or ECLIPSE.
You can find the installation documentation here: http://ant.apache.org/manual/install.html

## Problems
* Ensure that ANT_HOME is configured. You can check this using with
  * Windows:"echo %ANT_HOME%"
  * NiX: "echo $ANT_HOME"
  
# Download & compile

```
git clone git@github.com:g2x3k/Drftpd2Stable.git
cd Drftpd2Stable
./genkey.sh
ant
```

edit conf/* accordingly and launch your ftpd ...


CHANGELOG.txt
2007-04-08  tdsoul
* Port r1683 from trunk:
  
  java.io.File.list() may return null when the directory no longer
  exists or when an I/O error occurs. In either case, we have
  successfully deleted the file already so we can return as usual.
  This change makes sure we do not use a null pointer when list()
  returns null.
  
  Thanks to Dom for reporting this issue and helping to provide a
  root cause and potential fixes. Also thanks to zubov for his
  useful input.

2007-01-31  zubov
* commiting patch from djb61 for ticket #143 this will comply with
  the java api for Socket.setReceiveBufferSize()

2007-01-06  zubov
* updating MatchdirFilter to allow for slaves with - and + in the
  name, other places throughout the code may still error with this,
  use in caution
  
  Thanks xinu for pointing this out and the proposed fix This
  closes ticket #139

2007-01-06  fr0w
* src/net/sf/drftpd/master/command/plugins/LIST.java:
  Little casting problem while using NLST. Gotta hate Java 1.4
  codes :P

2007-01-05  zubov
* Closing extra file descriptors in, FtpConfig, ZipscriptConfig,
  CommandManagerFactory, SectionManager Adding a null file
  descriptor check before closing in SlaveManager, FileRemoteFile,
  and PropertyBuilder Adding comparisons to null Objects in
  .equals() for KeyedMap, RootCollection, and TransferIndex Fixing
  a call for statistics in AbstractUser, Monthly periods should
  return Monthly stats Fixing some other usages of null Objects
  where we know they are null, fixed for code clarity and no
  complaints from the IDE :)
  
  Thanks iKm!

2006-12-21  zubov
* fixed ConstantMirroring.isArchivedDir() method. It didn't
  properly handle the default case of all slaves being the
  destination

* fixed a buffer not being closed, thanks xinu

2006-12-16  zubov
* fixing #124, symlinks shouldn't be sent out to the slaves to be
  deleted since they don't reside there Thanks TDSoul for the
  proposed fix

2006-12-14  zubov
* added support for #137 site slave <slavename> unset <key>
  
  I've always thought this should be implemented as well
  
  This also fixes a small race condition that could have occured
  while setting/updating/removing a value

* fixed an almost certain concurrentmodification when cancelling
  transfers inside a directory

2006-12-13  tdsoul
* Update teflon's IMDB parser for compatability with iMDB website

2006-11-17  dark
* Cosmetic change to "site unnuke". The reason was not being
  trimmed which lead to leading whitespace. The "site nuke" command
  already takes care of this.

2006-10-23  tdsoul
* Minor enhancement ADDUSER/ADDIP/DELIP will now strip any commas 
  from ipmasks (for easy pasting)

2006-10-21  fr0w
* Fixed an error that should cause the slave to go offline if the
  Checksum method fail.

* Archive could not run as it should in DatedSections, it was only
  archiving the 'current' directory of the section instead of
  archiving the whole section.

2006-10-19  tdsoul
* Simultaneous upload/download of the same file (implemented in 
  r1158) failed to work properly on the Windows platform. If this
  change breaks the server on your platform, please advise.

2006-10-19  zubov
* Excellent write-up provided for a possibility of causing the
  ActiveConnection to be slow. In the case where dns servers are
  unresponsive, the lookup request could delay the connection long
  enough to disrupt the connection and create a timeout. The lookup
  was only done for human readable logging, so we'll just log the
  IPs instead. Thanks!
  
  http://www.drftpd.org/phpBB2/viewtopic.php?t=1238

2006-10-16  tdsoul
* SFVFile Interface updated -- Overloaded getTotalBytes() to allow
  it to return bytes for only completed files, not just those being
  transfered.
  
  Also added a getTotalBytesCompleted() convenience method.
  
  Default behavior is old behavior, return size of all files in the
  SFV, including those still being transfered.
  
  This change fixes a bug in the Sitebot which poorly calculated
  the expected size of a completed directory which contains an SFV
  file.
  
  Closes: #130

2006-10-06  tdsoul
* Clean up NukeLog check a bit, and fix some bugs that could be
  exploited to circumvent it.

2006/06/16 fr0w <>
* Add support for SSL slave-to-slave transfers issued by JobManager.

2006/06/03 dark <>
* Fixed: Nuke multiplier of 0 should not remove credits.  Thanks _Dink.
* Fixed: SiteBot output the full path rather than the "section-stripped" path on Nuke. Thanks JJLIT.
* Fixed: The incorrect checksum is reported when issuing "site rescan".
* Added: A 'totalsize' variable to the sitebot for announcing the total size of files in an SFV.
* Fixed: totalfiles variable to more accurately reflect reality on SITE PRE.
* Fixed: a bug with !kick, which was comparing seconds to milliseconds.
* Fixed: Time.formatTime() was being passed seconds, instead of the required milliseconds.

2006/05/08 Anonymous Coward <tdsoul@gmail.com>
* If ratio is 0, nuked amount should always return 0.

2006/04/29 dark <>
* Fixed an "off by one" bug in the max_sim code, which gave users an extra transfer over the max.
* Fixed another multi-announce issue
* DIZPlugin should only get zip files that are not currently transfering

2006/03/30 Matt Zimmerman <zubov@users.sourceforge.net>
* Introduced the FileStillTransferringException to resolve several issues related to sfv files.
* Fixed the incompatibility with tlswrap and brought drftpd into compliance with the tls-ftp spec
* Added imdb.syntax for outputing help when !imdb is not passed any parameters, thanks to JJLIT

2006/03/07 dark <>
* Fixed a bug related to merging existing files w/o any slaves already assigned to them.

2006/02/24 Matt Zimmerman <zubov@users.sourceforge.net>
* Fixed a possible deadlock issue when a transfer is aborted.
* Fixed wrong "idletime" sometimes being displayed by UserManagement.

2006/02/22 Matt Zimmerman <zubov@users.sourceforge.net>
* 2.0.3 released

src/net/sf/drftpd/util/MD5.java
/*
 * This file is part of DrFTPD, Distributed FTP Daemon.
 *
 * DrFTPD is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * DrFTPD is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with DrFTPD; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * MD5.java -- Class implementing the MD5 algorithm as specified in RFC1321.
 * Copyright (C) 1999, 2002 Free Software Foundation, Inc.
 * 
 * Originally package gnu.java.security.provider, extracted and moved
 * into com.graeff.pwencode by Tom Aeby for practical reasons.
 * 
 * This file is part of GNU Classpath.
 * 
 * GNU Classpath is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 * 
 * GNU Classpath is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with GNU Classpath; see the file COPYING.  If not, write to the
 * Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
 * 02111-1307 USA.
 * 
 * Linking this library statically or dynamically with other modules is
 * making a combined work based on this library.  Thus, the terms and
 * conditions of the GNU General Public License cover the whole
 * combination.
 * 
 * As a special exception, the copyright holders of this library give you
 * permission to link this library with independent modules to produce an
 * executable, regardless of the license terms of these independent
 * modules, and to copy and distribute the resulting executable under
 * terms of your choice, provided that you also meet, for each linked
 * independent module, the terms and conditions of the license of that
 * module.  An independent module is a module which is not derived from
 * or based on this library.  If you modify this library, you may extend
 * this exception to your version of the library, but you are not
 * obligated to do so.  If you do not wish to do so, delete this
 * exception statement from your version.
 */

package net.sf.drftpd.util;
import java.security.MessageDigest;

/**
 This class implements the MD5 algorithm as described in RFC1321.
 
 @see java.security.MessageDigest
 */
public class MD5 extends MessageDigest implements Cloneable
{
	private final int W[] = new int[16];
	private long bytecount;
	private int A;
	private int B;
	private int C;
	private int D;
	
	public MD5()
	{
		super("MD5");
		engineReset ();
	}
	
	public Object clone()
	{
		return new MD5 (this);
	}
	
	private MD5 (MD5 copy)
	{
		this ();
		bytecount = copy.bytecount;
		A = copy.A;
		B = copy.B;
		C = copy.C;
		D = copy.D;
		System.arraycopy (copy.W, 0, W, 0, 16);
	}
	
	public int engineGetDigestLength()
	{
		return 20;
	}
	
	// Intialize the A,B,C,D needed for the hash
	public void engineReset()
	{
		bytecount = 0;
		A = 0x67452301;
		B = 0xefcdab89;
		C = 0x98badcfe;
		D = 0x10325476;
		for(int i = 0; i < 16; i++)
			W[i] = 0;
	}
	
	public void engineUpdate (byte b)
	{
		int i = (int)bytecount % 64;
		int shift = (3 - i % 4) * 8;
		int idx = i / 4;
		
		// if you could index ints, this would be: W[idx][shift/8] = b
		W[idx] = (W[idx] & ~(0xff << shift)) | ((b & 0xff) << shift);
		
		// if we've filled up a block, then process it
		if ((++ bytecount) % 64 == 0)
			munch ();
	}
	
	public void engineUpdate (byte bytes[], int off, int len)
	{
		if (len < 0)
			throw new ArrayIndexOutOfBoundsException ();
		
		int end = off + len;
		while (off < end)
			engineUpdate (bytes[off++]);
	}
	
	public byte[] engineDigest()
	{
		long bitcount = bytecount * 8;
		engineUpdate ((byte)0x80); // 10000000 in binary; the start of the padding
		
		// add the rest of the padding to fill this block out, but leave 8
		// bytes to put in the original bytecount
		while ((int)bytecount % 64 != 56)
			engineUpdate ((byte)0);
		
		// add the length of the original, unpadded block to the end of
		// the padding
		W[14] = SWAP((int)(0xffffffff & bitcount));
		W[15] = SWAP((int)(0xffffffff & (bitcount >>> 32)));
		bytecount += 8;
		
		// digest the fully padded block
		munch ();
		
		A = SWAP(A);
		B = SWAP(B);
		C = SWAP(C);
		D = SWAP(D);
		byte[] result = new byte[] {(byte)(A >>> 24), (byte)(A >>> 16),
				(byte)(A >>> 8), (byte)A,
				(byte)(B >>> 24), (byte)(B >>> 16),
				(byte)(B >>> 8), (byte)B,
				(byte)(C >>> 24), (byte)(C >>> 16),
				(byte)(C >>> 8), (byte)C,
				(byte)(D >>> 24), (byte)(D >>> 16),
				(byte)(D >>> 8), (byte)D};
		
		engineReset ();
		return result;
	}
	
	private int F( int X, int Y, int Z)
	{
		return ((X & Y) | (~X & Z));
	}
	
	private int G( int X, int Y, int Z)
	{
		return ((X & Z) | (Y & ~Z));
	}
	
	private int H( int X, int Y, int Z)
	{
		return (X ^ Y ^ Z);
	}
	
	private int I( int X, int Y, int Z)
	{
		return (Y ^ (X | ~Z));
	}
	
	private int rotateLeft( int i, int count)
	{
		//Taken from FIPS 180-1
		return ( (i << count) | (i >>> (32 - count)) ) ;
	}
	
	/* Round 1. */
	private int FF( int a, int b, int c, int d, int k, int s, int i)
	{
		/* Let [abcd k s i] denote the operation */
		a += F(b,c,d) + k + i;
		return b + rotateLeft(a, s); 
	} 
	/* Round 2. */
	private int GG( int a, int b, int c, int d, int k, int s, int i)
	{
		/* Let [abcd k s i] denote the operation */
		a += G(b,c,d) + k + i;
		return b + rotateLeft(a, s);
	} 
	/* Round 3. */
	private int HH( int a, int b, int c, int d, int k, int s, int i)
	{
		/* Let [abcd k s t] denote the operation */
		a += H(b,c,d) + k + i;
		return b + rotateLeft(a, s);
	} 
	
	/* Round 4. */
	private int II( int a, int b, int c, int d, int k, int s, int i)
	{
		/* Let [abcd k s t] denote the operation */
		a += I(b,c,d) + k + i;
		return b + rotateLeft(a, s);
	} 
	
	private int SWAP(int n)
	{
		//Copied from md5.c in FSF Gnu Privacy Guard 0.9.2
		return (( (0xff & n) << 24) | ((n & 0xff00) << 8) | ((n >>> 8) & 0xff00) | (n >>> 24));
	}
	
	private void munch()
	{
		int AA,BB,CC,DD, j;
		int X[] = new int[16];
		
		/* Copy block i into X. */
		for(j = 0; j < 16; j++)
			X[j] = SWAP(W[j]);
		
		/* Save A as AA, B as BB, C as CC, and D as DD. */
		AA = A;
		BB = B;
		CC = C;
		DD = D;
		
		/* The hex constants are from md5.c 
		 in FSF Gnu Privacy Guard 0.9.2 */
		/* Round 1. */
		/* Let [abcd k s i] denote the operation
		 a = b + ((a + F(b,c,d) + X[k] + T[i]) <<< s). */
		/* Do the following 16 operations. */
		A = FF(A,B,C,D,  X[0],  7,  0xd76aa478);
		D = FF(D,A,B,C,  X[1], 12,  0xe8c7b756);
		C = FF(C,D,A,B,  X[2], 17,  0x242070db);
		B = FF(B,C,D,A,  X[3], 22,  0xc1bdceee);
		
		A = FF(A,B,C,D,  X[4],  7,  0xf57c0faf);
		D = FF(D,A,B,C,  X[5], 12,  0x4787c62a);
		C = FF(C,D,A,B,  X[6], 17,  0xa8304613);
		B = FF(B,C,D,A,  X[7], 22,  0xfd469501);
		
		A = FF(A,B,C,D,  X[8],  7,  0x698098d8);
		D = FF(D,A,B,C,  X[9], 12,  0x8b44f7af);
		C = FF(C,D,A,B, X[10], 17,  0xffff5bb1);
		B = FF(B,C,D,A, X[11], 22,  0x895cd7be);
		
		A = FF(A,B,C,D, X[12],  7,  0x6b901122);
		D = FF(D,A,B,C, X[13], 12,  0xfd987193);
		C = FF(C,D,A,B, X[14], 17,  0xa679438e);
		B = FF(B,C,D,A, X[15], 22,  0x49b40821);
		
		/* Round 2. */
		/* Let [abcd k s i] denote the operation
		 a = b + ((a + G(b,c,d) + X[k] + T[i]) <<< s). */
		/* Do the following 16 operations. */
		A = GG(A,B,C,D,  X[1],  5, 0xf61e2562);
		D = GG(D,A,B,C,  X[6],  9, 0xc040b340);
		C = GG(C,D,A,B, X[11], 14, 0x265e5a51);
		B = GG(B,C,D,A,  X[0], 20, 0xe9b6c7aa);
		
		A = GG(A,B,C,D,  X[5],  5, 0xd62f105d);
		D = GG(D,A,B,C, X[10],  9, 0x02441453);
		C = GG(C,D,A,B, X[15], 14, 0xd8a1e681);
		B = GG(B,C,D,A,  X[4], 20, 0xe7d3fbc8);
		
		A = GG(A,B,C,D,  X[9],  5, 0x21e1cde6);
		D = GG(D,A,B,C, X[14],  9, 0xc33707d6);
		C = GG(C,D,A,B,  X[3], 14, 0xf4d50d87);
		B = GG(B,C,D,A,  X[8], 20, 0x455a14ed);
		
		A = GG(A,B,C,D, X[13],  5, 0xa9e3e905);
		D = GG(D,A,B,C,  X[2],  9, 0xfcefa3f8);
		C = GG(C,D,A,B,  X[7], 14, 0x676f02d9);
		B = GG(B,C,D,A, X[12], 20, 0x8d2a4c8a);
		
		/* Round 3. */
		/* Let [abcd k s t] denote the operation
		 a = b + ((a + H(b,c,d) + X[k] + T[i]) <<< s). */
		/* Do the following 16 operations. */
		A = HH(A,B,C,D,  X[5],  4, 0xfffa3942);
		D = HH(D,A,B,C,  X[8], 11, 0x8771f681);
		C = HH(C,D,A,B, X[11], 16, 0x6d9d6122);
		B = HH(B,C,D,A, X[14], 23, 0xfde5380c);
		
		A = HH(A,B,C,D,  X[1],  4, 0xa4beea44);
		D = HH(D,A,B,C,  X[4], 11, 0x4bdecfa9);
		C = HH(C,D,A,B,  X[7], 16, 0xf6bb4b60);
		B = HH(B,C,D,A, X[10], 23, 0xbebfbc70);
		
		A = HH(A,B,C,D, X[13],  4, 0x289b7ec6);
		D = HH(D,A,B,C,  X[0], 11, 0xeaa127fa);
		C = HH(C,D,A,B,  X[3], 16, 0xd4ef3085);
		B = HH(B,C,D,A,  X[6], 23, 0x04881d05);
		
		A = HH(A,B,C,D,  X[9],  4, 0xd9d4d039);
		D = HH(D,A,B,C, X[12], 11, 0xe6db99e5);
		C = HH(C,D,A,B, X[15], 16, 0x1fa27cf8);
		B = HH(B,C,D,A,  X[2], 23, 0xc4ac5665);
		
		/* Round 4. */
		/* Let [abcd k s t] denote the operation
		 a = b + ((a + I(b,c,d) + X[k] + T[i]) <<< s). */
		/* Do the following 16 operations. */
		A = II(A,B,C,D,  X[0],  6, 0xf4292244);
		D = II(D,A,B,C,  X[7], 10, 0x432aff97);
		C = II(C,D,A,B, X[14], 15, 0xab9423a7);
		B = II(B,C,D,A,  X[5], 21, 0xfc93a039);
		
		A = II(A,B,C,D, X[12],  6, 0x655b59c3);
		D = II(D,A,B,C,  X[3], 10, 0x8f0ccc92);
		C = II(C,D,A,B, X[10], 15, 0xffeff47d);
		B = II(B,C,D,A,  X[1], 21, 0x85845dd1);
		
		A = II(A,B,C,D,  X[8],  6, 0x6fa87e4f);
		D = II(D,A,B,C, X[15], 10, 0xfe2ce6e0);
		C = II(C,D,A,B,  X[6], 15, 0xa3014314);
		B = II(B,C,D,A, X[13], 21, 0x4e0811a1);
		
		A = II(A,B,C,D,  X[4],  6, 0xf7537e82);
		D = II(D,A,B,C, X[11], 10, 0xbd3af235);
		C = II(C,D,A,B,  X[2], 15, 0x2ad7d2bb);
		B = II(B,C,D,A,  X[9], 21, 0xeb86d391);
		
		/* Then perform the following additions. (That is increment each
		 of the four registers by the value it had before this block
		 was started.) */
		A = A + AA;
		B = B + BB;
		C = C + CC;
		D = D + DD;
	}

}